File: gateway/src/main/java/com/beat/gateway/security/JwtAuthenticationFilter.java
package com.beat.gateway.security;

import com.beat.contracts.auth.JwtTokenPort;
import com.beat.contracts.auth.TokenValidationResult;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Collection;
import java.util.List;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import org.springframework.web.filter.OncePerRequestFilter;

@Slf4j
@Component
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {

	private static final String ROLE_ADMIN = "ROLE_ADMIN";
	private static final String ROLE_MEMBER = "ROLE_MEMBER";

	private final JwtTokenPort jwtTokenPort;

	@Override
	protected void doFilterInternal(
		@NonNull HttpServletRequest request,
		@NonNull HttpServletResponse response,
		@NonNull FilterChain filterChain
	) throws ServletException, IOException {

		final String token = getJwtFromRequest(request);

		if (!StringUtils.hasText(token)) {
			log.debug("JWT Token not found in request header. Assuming guest access or public API request.");
			filterChain.doFilter(request, response);
			return;
		}

		try {
			TokenValidationResult validationResult = jwtTokenPort.validateToken(token);

			if (validationResult != TokenValidationResult.VALID) {
				handleInvalidToken(validationResult, response);
				return;
			}

			setAuthentication(token, request);
			filterChain.doFilter(request, response);
		} catch (IllegalArgumentException exception) {
			log.warn("Invalid JWT claims: {}", exception.getMessage());
			response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
		} catch (Exception exception) {
			log.error("JWT Authentication Exception: ", exception);
			response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
		}
	}

	private void setAuthentication(String token, HttpServletRequest request) {
		Long memberId = jwtTokenPort.getMemberId(token);
		String roleName = jwtTokenPort.getRoleName(token);

		Collection<GrantedAuthority> authorities = List.of(new SimpleGrantedAuthority(roleName));
		UsernamePasswordAuthenticationToken authentication = createAuthentication(memberId, authorities, roleName);
		authentication.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));

		SecurityContextHolder.getContext().setAuthentication(authentication);
	}

	private void handleInvalidToken(TokenValidationResult validationResult, HttpServletResponse response) {
		if (validationResult == TokenValidationResult.EXPIRED) {
			response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
			return;
		}
		response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
	}

	private UsernamePasswordAuthenticationToken createAuthentication(
		Long memberId,
		Collection<GrantedAuthority> authorities,
		String roleName
	) {
		if (ROLE_ADMIN.equals(roleName)) {
			return new AdminAuthentication(memberId, null, authorities);
		}
		if (ROLE_MEMBER.equals(roleName)) {
			return new MemberAuthentication(memberId, null, authorities);
		}
		return new UsernamePasswordAuthenticationToken(memberId, null, authorities);
	}

	private String getJwtFromRequest(HttpServletRequest request) {
		String bearerToken = request.getHeader("Authorization");
		if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
			return bearerToken.substring("Bearer ".length());
		}
		return null;
	}
}

File: gateway/src/main/kotlin/com/beat/gateway/authentication/internal/JwtAuthenticationFilter.kt
package com.beat.gateway.authentication.internal

import com.beat.contracts.auth.JwtTokenPort
import com.beat.contracts.auth.JwtTokenType
import com.beat.contracts.auth.TokenValidationResult
import com.beat.observability.logging.filter.BaseMdcLoggingFilter
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.LoggerFactory
import org.slf4j.MDC
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component("gatewayJwtAuthenticationFilter")
class JwtAuthenticationFilter(
    private val jwtTokenPort: JwtTokenPort,
) : OncePerRequestFilter() {

    private val log = LoggerFactory.getLogger(JwtAuthenticationFilter::class.java)
    private val authenticationDetailsSource = WebAuthenticationDetailsSource()

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        val token = request.bearerToken()

        if (token == null) {
            log.debug("JWT Token not found in request header. Assuming guest access or public API request.")
            filterChain.doFilter(request, response)
            return
        }

        try {
            val validationResult = jwtTokenPort.validateAccessToken(token)

            if (validationResult != TokenValidationResult.VALID) {
                response.status = validationResult.toHttpStatus()
                return
            }

            authenticate(token, request)
            filterChain.doFilter(request, response)
        } catch (exception: IllegalArgumentException) {
            log.warn("Invalid JWT claims: {}", exception.message)
            response.status = HttpServletResponse.SC_UNAUTHORIZED
        } catch (exception: Exception) {
            log.error("JWT Authentication Exception: ", exception)
            response.status = HttpServletResponse.SC_INTERNAL_SERVER_ERROR
        }
    }

    private fun authenticate(token: String, request: HttpServletRequest) {
        val memberId = jwtTokenPort.getMemberId(token, JwtTokenType.ACCESS)
        val roleName = jwtTokenPort.getRoleName(token, JwtTokenType.ACCESS)

        val authentication = createAuthentication(memberId, roleName).apply {
            details = authenticationDetailsSource.buildDetails(request)
        }

        SecurityContextHolder.getContext().authentication = authentication
        MDC.put(BaseMdcLoggingFilter.USER_ID_KEY, memberId.toString())
    }

    private fun createAuthentication(memberId: Long, roleName: String): UsernamePasswordAuthenticationToken {
        val authorities: Collection<GrantedAuthority> = listOf(SimpleGrantedAuthority(roleName))
        return when (roleName) {
            ROLE_ADMIN -> AdminAuthentication(memberId, authorities)
            ROLE_MEMBER -> MemberAuthentication(memberId, authorities)
            else -> UsernamePasswordAuthenticationToken(memberId, null, authorities)
        }
    }

    /**
     * `EXPIRED -> 401`, 그 외 실패 -> `400`은 현재 client 호환 계약이다.
     */
    private fun TokenValidationResult.toHttpStatus(): Int = when (this) {
        TokenValidationResult.EXPIRED -> HttpServletResponse.SC_UNAUTHORIZED
        else -> HttpServletResponse.SC_BAD_REQUEST
    }

    private fun HttpServletRequest.bearerToken(): String? =
        getHeader(HEADER_AUTHORIZATION)
            ?.takeIf { it.startsWith(BEARER_PREFIX) }
            ?.removePrefix(BEARER_PREFIX)
            ?.takeIf(String::isNotBlank)

    companion object {
        private const val ROLE_ADMIN = "ROLE_ADMIN"
        private const val ROLE_MEMBER = "ROLE_MEMBER"
        private const val HEADER_AUTHORIZATION = "Authorization"
        private const val BEARER_PREFIX = "Bearer "
    }
}
